cmd/internal/kube/inject.go

```diff
@@ -0,0 +1,247 @@
+		}
+		// Append the injected YAML to the output
+		out.Write(rawInjected)
+
+		// If the user did not specify an output file, print the output to stdout
+		if injectOutputFlag == "" {
+			printer.Stdout.RawOutput(out.String())
+			return nil
+		}
+
+		// Write the output to the specified file
+		if err := writeFile(out.Bytes(), injectOutputFlag); err != nil {
+			return err
+		}
+		printer.Infof("Injected YAML written to %s\n", injectOutputFlag)
+
+		return nil
+	},
+	PersistentPreRun: func(cmd *cobra.Command, args []string) {
+		// Initialize the telemetry client, but do not allow any logs to be printed
+		telemetry.Init(false)
+	},
+}
+
+// A parsed representation of the `--secret` option.
+type secretGenerationOptions struct {
+	// Whether to inject a secret
+	ShouldInject bool
+	// The path to the secret file
+	Filepath optionals.Optional[string]
+}
+
+func createSidecar(projectName string) v1.Container {
+	sidecar := v1.Container{
+		Name:  "akita",
+		Image: "akitasoftware/cli:latest",
+		Env: []v1.EnvVar{
+			{
+				Name: "AKITA_API_KEY_ID",
+				ValueFrom: &v1.EnvVarSource{
+					SecretKeyRef: &v1.SecretKeySelector{
+						LocalObjectReference: v1.LocalObjectReference{
+							Name: "akita-secrets",
+						},
+						Key: "akita-api-key",
+					},
+				},
+			},
+			{
+				Name: "AKITA_API_KEY_SECRET",
+				ValueFrom: &v1.EnvVarSource{
+					SecretKeyRef: &v1.SecretKeySelector{
+						LocalObjectReference: v1.LocalObjectReference{
+							Name: "akita-secrets",
+						},
+						Key: "akita-api-secret",
+					},
+				},
+			},
+		},
+		Lifecycle: &v1.Lifecycle{
+			PreStop: &v1.LifecycleHandler{
+				Exec: &v1.ExecAction{
+					Command: []string{
+						"/bin/sh",
+						"-c",
+						"AKITA_PID=$(pgrep akita) && kill -2 $AKITA_PID && tail -f /proc/$AKITA_PID/fd/1",
+					},
+				},
+			},
+		},
+		Args: []string{"apidump", "--project", projectName},
+		SecurityContext: &v1.SecurityContext{
+			Capabilities: &v1.Capabilities{Add: []v1.Capability{"NET_RAW"}},
+		},
+	}
+
+	return sidecar
+}
+
+// Parses the given value for the `--secret` option.
+func resolveSecretGenerationOptions(flagValue string) secretGenerationOptions {
+	if flagValue == "" || flagValue == "false" {
+		return secretGenerationOptions{
+			ShouldInject: false,
+			Filepath:     optionals.None[string](),
+		}
+	}
+
+	if flagValue == "true" {
+		return secretGenerationOptions{
+			ShouldInject: true,
+			Filepath:     optionals.None[string](),
+		}
+	}
+
+	return secretGenerationOptions{
+		ShouldInject: true,
+		Filepath:     optionals.Some(flagValue),
+	}
+}
+
+func init() {
+	injectCmd.Flags().StringVarP(
+		&injectFileNameFlag,
+		"file",
+		"f",
+		"",
+		"Path to the Kubernetes YAML file to be injected. This should contain a Deployment object.",
+	)
+	_ = injectCmd.MarkFlagRequired("file")
+
+	injectCmd.Flags().StringVarP(
+		&injectOutputFlag,
+		"output",
+		"o",
+		"",
+		"Path to the output file. If not specified, the output will be printed to stdout.",
+	)
```

Comment:
What if we named these options `--in` and `--out` instead? I just caught myself thinking that it's `--out`, because that's what I'm used to with other command-line tools.

Reply:
We are trying to match the conventions in the K8S ecosystem.

Istioctl uses
--output <string> | -o | Modified output Kubernetes resource filename (default ``)
--filename <string> | -f | Input Kubernetes resource filename (default ``)

Linkerd doesn't allow specifying at all, it's stdin/stdout or nothing.

nginx-meshctl also only allows stdin/stdout as far as I can tell

Kubectl itself, unfortunately, uses -o / --output but it's for the format, not the filename, but -f / --filename for input and output file parameters.

So I think -f/--filename for input is probably the rough consensus, as far as there can be said to be one.

Reply:
Really the core use case here is

`akita kube inject -f deployment.yaml | kubectl apply -f -` 

and everything else is gravy :)

Reply:
Ah, I see. Makes sense. Resolving.